### PR TITLE
Simple range proof verification

### PR DESCRIPTION
**Pull request by rkuris**
## Why this should be merged

Partial verification of range proofs, without generating tries and checking hashes.

## How this works

The current implementation provides a foundation for range proof verification by:

* Checking structural validity (key ordering, non-empty proof)
* Validating that the requested range matches the provided proof
* Cryptographically verifying the start and end boundary proofs

What's NOT included (tracked in #738):

* Full trie reconstruction from the proof
* Verification of intermediate nodes between boundaries
* Complete validation that no keys are missing within the proven range

## How this was tested

The test suite in range.rs, which includes tests for:

* Valid range proofs with various configurations
* Invalid proofs (modified keys, values, gaps, etc.)
* Edge cases (single element, empty ranges, full trie)

Note: Most tests are currently #ignore because they are not checked yet and will be fixed in followup PRs.

### CODE REVIEW

File: firewood/src/lib.rs [hunk 1]
```diff
@@ -175,7 +175,7 @@ pub use firewood_storage::logger;
 /// In the event of unexpected behavior in testing, disable this first before
 /// looking elsewhere.
 fn init_logger() {
-    env_logger::Builder::from_env(env_logger::Env::default().default_filter_or("trace"))
+    env_logger::Builder::from_env(env_logger::Env::default().default_filter_or("info"))
```
Comment (rkuris) [inline firewood/src/lib.rs:184]: trace logging for all tests is very expensive, noticed a big improvement here.

Sorry to sneak this in to this PR, really it belongs as a separate PR. Complain if you agree and I'll move it.

File: firewood/src/merkle/tests/range.rs [hunk 2]
```diff
@@ -104,10 +102,12 @@ fn test_bad_range_proof() {
             0 => {
                 // Modified key
                 keys[index] = rng.random::<[u8; 32]>(); // In theory it can't be same
+                continue 'skip_test;
```
[the remark this answers is not in the document]
Comment (rkuris) [reply firewood/src/merkle/tests/range.rs]: The whole point of this PR is to enable the parts of the test that do work, and out of order is an integral part of this test. The others will get skipped.

File: firewood/src/merkle/mod.rs [hunk 3]
```diff
@@ -259,12 +259,64 @@ impl<T: TrieReader> Merkle<T> {
...
+        {
+            return Err(api::Error::InvalidRange {
+                start_key: requested_first.as_ref().to_vec().into(),
+                end_key: left.0.as_ref().to_vec().into(),
+            });
+        }
+
+        proof
+            .start_proof()
+            .verify(&left.0, Some(&left.1), root_hash)?;
```
Comment (demosdemon) [inline firewood/src/merkle/mod.rs]: This is incorrect. When we generate proofs, we always generate a proof for the lower bound. The proof for the lower bound may or may not prove the first key provided in the range. If the lower bound does not exist, the provided proof may not have all the information required to prove the inclusion of the first key.

File: firewood/src/merkle/mod.rs [hunk 4]
```diff
@@ -259,12 +259,64 @@ impl<T: TrieReader> Merkle<T> {
...
+        {
+            return Err(api::Error::InvalidRange {
+                start_key: right.0.as_ref().to_vec().into(),
+                end_key: requested_last.as_ref().to_vec().into(),
+            });
+        }
+
+        proof
+            .end_proof()
+            .verify(&right.0, Some(&right.1), root_hash)?;
```
Comment (demosdemon) [inline firewood/src/merkle/mod.rs]: Similarly here, when we generate the upper bound proof, we generate a proof for the final key in the yielded range only if we truncated the range. Otherwise, we generate a proof for the requested upper bound. This proof may or may not be sufficient to verify the final yielded key.

File: firewood/src/merkle/mod.rs [hunk 5]
```diff
@@ -259,12 +259,64 @@ impl<T: TrieReader> Merkle<T> {
...
+        {
+            return Err(api::Error::ProofError(
+                ProofError::NonMonotonicIncreaseRange,
+            ));
+        }
+
+        // check that the start and end proofs are valid
+        let left = key_values
+            .first()
+            .ok_or(api::Error::ProofError(ProofError::Empty))?;
```
Comment (demosdemon) [inline firewood/src/merkle/mod.rs]: This is incorrect. An empty set of key-values is only invalid if both key bounds are empty. An empty set of key-values with a bounded range correctly represents an empty set of key-values between the two keys.

File: firewood/src/merkle/mod.rs (hunk 3 above)
Comment (rkuris) [reply firewood/src/merkle/mod.rs]: Seems like we should perhaps construct the key provided and make sure its <= to the first key.

File: firewood/src/merkle/tests/range.rs [hunk 6]
```diff
@@ -69,7 +68,6 @@ fn test_range_proof() {
 #[test]
 // Tests a few cases which the proof is wrong.
 // The prover is expected to detect the error.
-#[ignore = "https://github.com/ava-labs/firewood/issues/738"]
 fn test_bad_range_proof() {
```
Comment (RodrigoVillar) [inline firewood/src/merkle/tests/range.rs]: (q / feel-free-to-ignore): I wonder if it's cleaner to split this test into two - one test that tests with verification and one test that doesn't. Having the `'skip_test` loop label is hard to follow considering that we also have `continue` statements which should be kept around.

File: firewood/src/merkle/mod.rs [hunk 7]
```diff
@@ -262,12 +262,64 @@ impl<T: TrieReader> Merkle<T> {
...
-        _proof: &RangeProof<impl KeyType, impl ValueType, impl ProofCollection>,
+        first_key: Option<impl KeyType>,
+        last_key: Option<impl KeyType>,
+        root_hash: &TrieHash,
+        proof: &RangeProof<impl KeyType, impl ValueType, impl ProofCollection>,
     ) -> Result<(), api::Error> {
-        todo!()
+        // check that the keys are in ascending order
+        let key_values = proof.key_values();
+        if !key_values
```
Comment (bernard-avalabs) [inline firewood/src/merkle/mod.rs:256]: nit: Using is_sorted_by is slightly simpler:

```
        if !key_values
            .iter()
            .is_sorted_by(|a, b| a.0.as_ref() < b.0.as_ref())
```

File: firewood/src/merkle/mod.rs (hunk 3 above)
Comment (rkuris) [reply firewood/src/merkle/mod.rs]: This should be resolved with the changes I made.

File: firewood/src/merkle/mod.rs (hunk 5 above)
Comment (rkuris) [reply firewood/src/merkle/mod.rs]: Fixed.

File: firewood/src/merkle/mod.rs [hunk 8]
```diff
@@ -262,12 +262,131 @@ impl<T: TrieReader> Merkle<T> {
...
+                root_hash,
+            )?;
+
+            // If the requested lower bound differs from the first yielded key,
+            // the boundary proof should still be able to prove that yielded key.
+            // This rejects range proofs that skip keys between the bound and payload.
+            if let Some((left_key, _left_value)) = left
+                && requested_first.as_ref() != left_key.as_ref()
+            {
+                let mut gap_iter = self
```
Comment (bernard-avalabs) [inline firewood/src/merkle/mod.rs]: Clarification question:

Is this trying to verify that there is no gap between the end of the last range proof partition that was applied and the start of this range proof partition? 

What if the original trie that the range proofs were being applied to was no empty? Do range proofs include deletes for its range to ensure that existing keys are deleted?

File: firewood/src/merkle/mod.rs [hunk 9]
```diff
@@ -262,12 +262,131 @@ impl<T: TrieReader> Merkle<T> {
...
+                    .key_value_iter_from_key(requested_first.as_ref())
+                    .stop_after_key(Some(left_key.as_ref()));
+
```
Comment (demosdemon) [inline firewood/src/merkle/mod.rs]: This doesn't really have anything to do with verification and is conflating local state with the proof state. This only works if the `self` trie already contains all of the keys and values in the range proof. If it doesn't, which is typical when verifying a range proof before applying, this will fail.

File: firewood/src/merkle/mod.rs [hunk 10]
```diff
@@ -262,12 +262,131 @@ impl<T: TrieReader> Merkle<T> {
...
+                .end_proof()
+                .verify(requested_last.as_ref(), expected_end_value, root_hash)?;
+
+            // Symmetric boundary-to-edge consistency check for the upper bound.
+            if let Some((right_key, _right_value)) = right
+                && requested_last.as_ref() != right_key.as_ref()
+            {
+                let mut gap_iter = self
+                    .key_value_iter_from_key(right_key.as_ref())
+                    .stop_after_key(Some(requested_last.as_ref()));
```
Comment (demosdemon) [inline firewood/src/merkle/mod.rs]: same here, checking the `self` trie will only yield the expected result if the local trie already has the same state as the proof

File: firewood/src/merkle/mod.rs (hunk 8 above)
Comment (rkuris) [reply firewood/src/merkle/mod.rs]: No. Range proofs do NOT contain deletes. It is assumed that all values between the two endpoints specified in the range proof contain no other values. If there are any values in that range other than the ones specified in the proof, they are not in the target revision, and should be removed implicitly if this range proof is being applied.

Unlike a change proof, since there is no "base" revision for a range proof, we don't know what keys and/or values might be there. We only know what's in the version being asked about.

File: firewood/src/merkle/mod.rs [hunk 11]
```diff
@@ -262,12 +262,131 @@ impl<T: TrieReader> Merkle<T> {
...
+
+        if key_values.is_empty() && first_key.is_none() && last_key.is_none() {
+            return Err(api::Error::ProofError(ProofError::Empty));
+        }
+
+        let left = key_values.first();
+        let right = key_values.last();
+
+        // Verify that first_key (if provided) is <= the first key in the proof
+        if let (Some(ref requested_first), Some((left_key, _))) = (first_key.as_ref(), left)
```
Comment (alarso16) [inline firewood/src/merkle/mod.rs]: I _think_ this `ref` is unnecessary (also below)

File: firewood/src/merkle/mod.rs (hunk 9 above)
Comment (rkuris) [reply firewood/src/merkle/mod.rs]: Removed key_value_iter_from_key gap checks in 5379b1b077faecd6966b025cfd8556a29a624b25

File: firewood/src/merkle/mod.rs (hunk 10 above)
Comment (rkuris) [reply firewood/src/merkle/mod.rs]: fixed in 5379b1b077faecd6966b025cfd8556a29a624b25

File: firewood/src/merkle/mod.rs (hunk 11 above)
Comment (rkuris) [reply firewood/src/merkle/mod.rs]: fixed all 4 in 21c9a65b54aad5aa849b08197946dd84e4b62a71

File: firewood/src/merkle/mod.rs [hunk 12]
```diff
@@ -262,12 +262,131 @@ impl<T: TrieReader> Merkle<T> {
...
+                        ProofError::RangeProofStartBeyondFirstKey,
+                    ));
+                };
+
+                if actual_first.as_ref() != left_key.as_ref() {
+                    return Err(api::Error::ProofError(
+                        ProofError::RangeProofStartBeyondFirstKey,
+                    ));
+                }
+            }
```
[the remark this answers is not in the document]
Comment (rkuris) [reply firewood/src/merkle/mod.rs]: fixed in 5379b1b077faecd6966b025cfd8556a29a624b25

File: firewood/src/merkle/mod.rs [hunk 13]
```diff
@@ -262,12 +262,131 @@ impl<T: TrieReader> Merkle<T> {
...
+                    return Err(api::Error::ProofError(
+                        ProofError::RangeProofStartBeyondFirstKey,
+                    ));
+                };
+
+                if actual_first.as_ref() != left_key.as_ref() {
+                    return Err(api::Error::ProofError(
+                        ProofError::RangeProofStartBeyondFirstKey,
+                    ));
+                }
```
[the remark this answers is not in the document]
Comment (rkuris) [reply firewood/src/merkle/mod.rs]: fixed in 5379b1b077faecd6966b025cfd8556a29a624b25
